### DISCUSSION TITLE
What is the established set of communities for the Karate Graph and the associated modularity value for that set of communities?

### DISCUSSION THREAD

**Discussion by JoanSano** [networkx/networkx#6418] [Q&A]
Hi all,

so me and my colleagues have recently come across an interesting observation. To be completely fair, a "reviewer 2" suggested an experiment that triggered this. We have not come up with a satisfactory solution to the question above, but we also think that "reviewer 2" will satisfied so this is just for my own piece of mind. 

The issues concerns the Louvain Community Detection Algorithm or LCDA [original reference](https://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008/meta?casa_token=NOaLRpNwMX0AAAAA:6KOcsMRD20q24GejU8npNdAoPy8jXHPzbxMa6AR7Xx1BR5b1HizxlemZ8-5d1YTmSFyP6otn73U) and the results for the Zachary-Karate club. The original reference as well as these ones [second reference](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.72.027104), [third reference](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0227538) provide a modularity of approximately 0.42 and a total of 4 communities. 

When computing LCDA and the resulting modularity we get 0.44:

```
lcda = louvain_communities(G, seed=None, resolution=1)
mod = modularity(G,communities_lcda[i])
```

We are aware that the node ordering can affect the results, so we tested for different seed as well as for different runs:

```
import networkx as nx
from networkx.algorithms.community import louvain_communities, modularity
import numpy as np
from joblib import Parallel, delayed
G = nx.karate_club_graph()

def seed_run(seed=None, runs=100):
  communities_lcda = []
  results_lcda = []
  Ncomms = 0
  for i in range(runs):
    communities_lcda.append(louvain_communities(G, seed=seed, resolution=1))
    results_lcda.append(modularity(G,communities_lcda[i]))
    Ncomms += len(communities_lcda[i])/100
  results_lcda = np.array(results_lcda)
  return results_lcda.mean(), results_lcda.std()/np.sqrt(runs)

results = Parallel(n_jobs=4)(delayed(seed_run)(seed, runs=100) for seed in range(10000))
```

![image](https://user-images.githubusercontent.com/95430406/217784077-81ad5875-9535-4247-a6bb-f10ca809fe7c.png)

For some seeds the implementation seems to get the 0.419... that is reported everywhere, but most of the results are in fact quite far from this value. Even more, the mean and standard deviation in both cases is approximately 0.44... +- 0.00...

```
import time
communities_lcda = []
results_lcda = []
Ncomms = 0
time_lcda = []
for i in range(10000):
  start = time.time()
  lcda = louvain_communities(G, seed=None, resolution=1)
  end = time.time()
  total_time = end - start

  time_lcda .append(total_time)
  communities_lcda.append(lcda)
  results_lcda.append(modularity(G,communities_lcda[i]))
  Ncomms += len(communities_lcda[i])/10000

results_lcda = np.array(results_lcda)
time_lcda = np.array(time_lcda)
```

![image](https://user-images.githubusercontent.com/95430406/217784230-92fdefbe-1ef7-428a-b347-a976b5fbe7db.png)

Letting the seed to be random returns very similar results, which is good.

We don't understand this. What is happending in NetworkX that no one else is talking about? Or we are missing some recent works on LCDA and new initialization strategies that actually return the correct modularity and partition. Just to provide a bit more context we are talking about new strategies of community detection with quantum methods so this might end up being something relevant.

I asked this question here as well, hoping that someone would answer with one or two equations but so far no luck. 
https://math.stackexchange.com/questions/4635046/what-is-the-stablished-value-for-the-modularity-of-the-karate-club?noredirect=1#comment9780684_4635046

Thanks in advance for any idea or comment about where we are wrong.
Best,
J

**Comment by dschult** [comment]
Let me see if I understand what you are saying:
Publications report that the modularity of the karate graph is 0.42, but the NetworkX implementation reports values ranging from 0.419 to 0.445 with a mean of 0.44 +/- 0.00.  Of course, the value of the modularity depends on what communities you provide to the modularity function. And the Louvain community detection algorithm is heuristic and depends on the node order which (in NetworkX) is determined by shuffling the nodes using a random seed. In NetworkX you show that some of the computed values agree with the publications, but the bulk of the results are about 5% higher.

I could imagine two questions arising from this exploration: 1) is NetworkX computing the modularity correctly? and 2) is NetworkX computing the Louvain communities correctly? I believe the answer to the first question is "yes". The modularity code is relatively simple and is deterministic (and not heuristic). It's always possible that a bug is there, but I'm pretty confident of that function.  The second question is harder to answer, because the Louvain algorithm is heuristic and involves choice of random values. 

The title of this discussion says "What is the established value for the modularity of the Karate Graph?". But that is an incomplete question. It should read: "What is the established set of communities for the Karate Graph and the associated modularity value for that set of communities?"  Your followup investigation then explores the variation in the sets of communities found by the NetworkX louvain implementation.

I think it is hard to compare the NetworkX variation in communities found with those of the papers because they don't report the variation in their results due to node order. Do you have any ideas about how to explore that?  It would be nice to use the implementation of the original authors to measure the variation in reported communities using their implementation. That would be something we could compare to.  It also might be good/easier to focus on the communities reported rather than the modularity value. At least I have this question: Are all of the modularity values of 0.419 using the same set of communities? or do multiple sets of communities result in the same modularity value?  It is really the set of communities that we are interested in (in my view). 

I am not aware of any new or different implementation details in the NetworkX code that is not in the original paper. It is always possible that there is a bug in the code, but I know of no repository of established communities for specific graphs to check with. That would be hard since Louvain is inherently heuristic and not deterministic. And finding communities is a combinatorially difficult process so finding the ground truth is usually intractable.  And of course, I might be missing something in my understanding.

**Reply by JoanSano** [reply]
Hi, 
Sorry for my late reply and many thanks for the answer. 

> Publications report that the modularity of the karate graph is 0.42, but the NetworkX implementation reports values ranging from 0.419 to 0.445 with a mean of 0.44 +/- 0.00. Of course, the value of the modularity depends on what communities you provide to the modularity function. And the Louvain community detection algorithm is heuristic and depends on the node order which (in NetworkX) is determined by shuffling the nodes using a random seed. In NetworkX you show that some of the computed values agree with the publications, but the bulk of the results are about 5% higher.

Indeed you understood my question and provided an excellent overview. 

In regards to possible bugs present in NetworkX:
> I could imagine two questions arising from this exploration: 1) is NetworkX computing the modularity correctly? and 2) is NetworkX computing the Louvain communities correctly? I believe the answer to the first question is "yes". The modularity code is relatively simple and is deterministic (and not heuristic). It's always possible that a bug is there, but I'm pretty confident of that function. The second question is harder to answer, because the Louvain algorithm is heuristic and involves choice of random values.

I agree with you that the implementation of the modularity function is not likely to be faulty. Furthermore, I also agree that the louvain method must have been checked enough times to be sufficiently safe. Yet, as you pointed out, the question remains.

In regards to the title of the discussion:
> The title of this discussion says "What is the established value for the modularity of the Karate Graph?". But that is an incomplete question. It should read: "What is the established set of communities for the Karate Graph and the associated modularity value for that set of communities?" Your followup investigation then explores the variation in the sets of communities found by the NetworkX louvain implementation.

I changed the title of the disucssion to what you proposed. 

In regards to this:
> I think it is hard to compare the NetworkX variation in communities found with those of the papers because they don't report the variation in their results due to node order. Do you have any ideas about how to explore that? It would be nice to use the implementation of the original authors to measure the variation in reported communities using their implementation. That would be something we could compare to. It also might be good/easier to focus on the communities reported rather than the modularity value. At least I have this question: Are all of the modularity values of 0.419 using the same set of communities? or do multiple sets of communities result in the same modularity value? It is really the set of communities that we are interested in (in my view).

However, In my opinion, the resulting modularity is as important as the set of communities from which it emerges. Modularity is often considered as the goodness of the result precisely due to the lack of ground truth for the resulting sets of communities [1,2]. I agree that the best thing to do would be to check the original implementatios of the algorithms. I will try to do this, but it may take me a bit of time since I will have to do this "off-record" (i.e., in my own spare time). Yet, for these purposes I wrote this small function that attemps to rapidly check your inquire. Short answer: Yes, it is using (or finding) the same set of communities. Here is the link to the google colab: https://colab.research.google.com/drive/1-gbUsosfT6yYxYVlDqCLsy6aL58Ij900#scrollTo=2hhMlWjnn0cr

In regards to this:
> I am not aware of any new or different implementation details in the NetworkX code that is not in the original paper. It is always possible that there is a bug in the code, but I know of no repository of established communities for specific graphs to check with. That would be hard since Louvain is inherently heuristic and not deterministic. And finding communities is a combinatorially difficult process so finding the ground truth is usually intractable. And of course, I might be missing something in my understanding.

I found that the documentation is actually citing an improvement on LCDA [2]. Can it be that the implemented algorithm already has this slightly advantageous characteristics? Of course it's not possible to check, because sadly the paper does not test its method on the karate club graph.

Actually, you have pointed out something very true, which is the lack of a rigorously stablished ground truth for the karate club or any other network. This network is always or nearly always taken as benchmark. I don't knkow if people would be interested in having a well stablished benchmark on this and maybe some other canonical networks. So I guess that the question you ask about me having any idea of "how to do this", would be answered by "maybe we need to keep rigorously benchmarking as many algorithms as possible for community detection". This should include not only modularity values but also, as you pointed out, the found communities. My question is if this would be of interest to anyone.

I am aware this might be a rather unusual thread to ask here, so I appreciate your answers and discussion.
Best,
J

[1] [https://journals.aps.org/pre/abstract/10.1103/PhysRevE.72.027104](url)
[2] [https://www.nature.com/articles/s41598-019-41695-z](url)
